ships/0026-shipwright-managed-push.md

```diff
@@ -0,0 +1,149 @@
+<!--
+Copyright The Shipwright Contributors
+
+SPDX-License-Identifier: Apache-2.0
+-->
+
+---
+title: shipwright-managed-push
+authors:
+  - @SaschaSchwarze0
+reviewers:
+  - @HeavyWombat
+  - @GabeMontero
+  - @ImJasonH
+approvers:
+  - @adambkaplan
+creation-date: 2022-01-19
+last-updated: 2022-01-19
+status: implementable
+---
+
+# Make the push operation to the container registry a shipwright-managed operation
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [x] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [docs](/docs/)
+
+## Open Questions [optional]
+
+None
+
+## Summary
+
+This ship describes why and how we provide means for build strategies to store the produced container image as a tarball in ephemeral storage instead of doing the image push. The push will then be performed by a shipwright-managed step.
+
+## Motivation
+
+Shipwrights mission is to turn source code into container images and push them to a container registry. So far, every build strategy author is responsible to implement the push operation. In most cases this is straight-forward because all of the tools that we use in our sample build strategies support this.
+
+It became more complex when we introduced results to pass information about the image to the BuildRun status: the digest and the size. In many cases, additional script-style logic had to be added to build strategies to parse standardized or propriatary files to access this information, and then to write it to response files. This made those strategies more complex.
```

Comment:
I am adding some references about how much was necessary to be more convincing.

(1) I think this can be compared with sources. The build strategy author has no flexibility here. We as Shipwright handle all this. This allows us to add more features for sources like the recently added local sources, without that build strategies authors must change anything. Just like loading sources, pushing the image is today part of any buildrun. I therefore do not see flexbility today. Every strategy basically implements the image push and the results magic. And a build strategy that does not implement a push "violates" the purpose of builds where users must define the output image with the semantic that the build output should be pushed there.
What we will do here is enabling the build strategy authors to focus on what they should focus = turning source code into a container image.

(2) Just related to simplify the results handling ... back at the time when we were using Tekton resources, I think we learned that this is complicated. The Tekton image resource expects an OCI image layout JSON file to extract the digest. But, only Kaniko and ko were supporting to write that file. This was indeed easy for them. But, for other tools that write other result files, this was not easily possible. We had a hack for Buildpacks, I think only downstream.

Reply:
I read the updates, they look good. No further comments, I concur here.